### PR TITLE
[IMP] core: allow adhoc fields in read_group

### PR DESCRIPTION
**Pull request by yelizariev**
There are few possible reasons why a computable field cannot be storable:

1. it depends on context:
   * current time
   * settings of a user (e.g. timezone)
   * a variable in ``env.context``

2. a field is used too rare to be stored in database (e.g. related field to
``partner_id.country_id.name``)

3. it would trigger too many/often recomputations

To allow grouping in such exceptional cases, this commit introduces *adhoc*
fields, which don't need to be specified in model definition as regular fields.

---

As a first application of the new features, this commit also replace
read_progress_bar with read_group, which is not just refactoring but a
performance improvement too.

Test demo crm.lead records:

```
| measurement        | before | after |
|--------------------+--------+-------|
| number of queries  |     10 |     3 |
| query time, ms     |     10 |     5 |
| remaining time, ms |     35 |     9 |

Test on 20K crm.lead records:

| measurement        | before | after |
|--------------------+--------+-------|
| number of queries  |    184 |     3 |
| query time, ms     |    900 |   165 |
| remaining time, ms |   5500 |    12 |
```

---

task-1915411
task-2155109
opw-2346901
closes #29292
closes #41421
closes #65136
closes #65854

### CODE REVIEW

**Comment by yelizariev** [comment]
Use-case in EE: https://github.com/odoo/enterprise/pull/16639

**Comment by yelizariev** [comment]
Test read_progress_bar (old) vs read_group (new) in CRM dashhoard

on demo database:
before: 
* number of quiries: 10
* query time: 8-12 ms
* remaining time: 30-40 ms

after:
 
* number of quiries: 3
* query time: 4-6 ms
* remaining time: 8-10 ms

on populated database (20 K crm.lead records with 3 mail.activity on each)

before

* number of quiries: 184
* query time: 850-1000 ms
* remaining time: 5000-6000

after:
 
* number of quiries: 3
* query time: 140-190 ms
* remaining time: 5-18 ms

File: odoo/fields.py [hunk 1]
```diff
@@ -691,11 +692,11 @@ def get_description(self, env):
 
     @property
     def _description_searchable(self):
-        return bool(self.store or self.search)
+        return bool(self.store or self.search or self.adhoc)
 
     @property
     def _description_sortable(self):
-        return (self.column_type and self.store) or (self.inherited and self.related_field._description_sortable)
+        return (self.column_type and self.store) or (self.inherited and self.related_field._description_sortable) or self.adhoc
```
Comment (yelizariev) [inline odoo/fields.py:699]: adhoc is the only change is fields definition. Potentially could be avoided by using ``_generate_adhoc_fields``

File: odoo/models.py [hunk 2]
```diff
@@ -69,7 +69,7 @@
 _schema = logging.getLogger(__name__ + '.schema')
 _unlink = logging.getLogger(__name__ + '.unlink')
 
-regex_order = re.compile('^(\s*([a-z0-9:_]+|"[a-z0-9:_]+")(\s+(desc|asc))?\s*(,|$))+(?<!,)$', re.I)
+regex_order = re.compile(r'^(\s*([a-z0-9:_.]+|"[a-z0-9:_.]+")(\s+(desc|asc))?\s*(,|$))+(?<!,)$', re.I)
```
Comment (yelizariev) [inline odoo/models.py:72]: Allows order by related fields

File: odoo/models.py [hunk 3]
```diff
@@ -4226,9 +4303,8 @@ def _load_records(self, data_list, update=False):
 
         return original_self.concat(*(data['record'] for data in data_list))
 
-    # TODO: ameliorer avec NULL
```
Comment (yelizariev) [inline odoo/models.py:4229]: It's very old comment and nobody  really knows what does it mean

File: odoo/models.py [hunk 4]
```diff
@@ -4490,7 +4570,8 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
         # the flush must be done before the _where_calc(), as the latter can do some selects
         self._flush_search(args, order=order)
 
-        query = self._where_calc(args)
+        adhoc = self._generate_adhoc_fields()
+        query = self._where_calc(args, adhoc=adhoc)
```
Comment (yelizariev) [inline odoo/models.py:4573]: Updating ``_search`` is needed to make Group By work in UI

File: odoo/osv/query.py [hunk 5]
```diff
@@ -103,11 +118,18 @@ def _join(self, kind, lhs_alias, lhs_column, rhs_table, rhs_column, link, extra=
         rhs_alias = _generate_table_alias(lhs_alias, link)
         assert rhs_alias not in self._tables, "Alias %r already in %s" % (rhs_alias, str(self))
 
+        assert lhs_column and rhs_column or extra, "Either column names or extra condition must be specified"
```
Comment (yelizariev) [inline odoo/osv/query.py:121]: See https://github.com/odoo/enterprise/pull/16639 for use case

**Review by yelizariev — COMMENTED** [review]
To validate names:

* adhoc
* query.get_sql_with()
* _generate_adhoc_fields

Maybe we can attach *adhoc fields* to model somehow in order to avoid adding optional adhoc parameter to every group by method

**Comment by odony** [comment]
as this PR touches a lot of sensitive files, it would be great to have some passes of reviews, e.g. with @rco-odoo, before taking it out of draft, no?

**Comment by odony** [comment]
quick comment: I'm not sure I see any conceptual improvement vs the previous attempt, most of the problems seem to be still present: developers need to implement low-level SQL methods, e.g. `_generate_adhoc_fields()`.

Tbh, I don't see much improvement over the various attempts that have already been done of overriding `read_group()` on a per-model basis to support random virtual fields?

I thought the idea was to automatically support grouping by related fields only, which should not require *any* extra work for the developers?

PS: The "group by follow-up level" idea is *not* something we want to support, it's way too exotic to be a core feature of the framework. I don't think we want to have it at all as a feature, but if we do, any other option sounds better than trying to support a native `group by (arbitrary expression)` (even a dumb cron job).

File: odoo/models.py [hunk 6]
```diff
@@ -2261,13 +2277,46 @@ def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=Fals
...
         """
-        result = self._read_group_raw(domain, fields, groupby, offset=offset, limit=limit, orderby=orderby, lazy=lazy)
-
         groupby = [groupby] if isinstance(groupby, str) else list(OrderedSet(groupby))
-        dt = [
-            f for f in groupby
-            if self._fields[f.split(':')[0]].type in ('date', 'datetime')    # e.g. 'date:month'
-        ]
+
+        adhoc = self._generate_adhoc_fields()
```
Comment (yelizariev) [inline odoo/models.py:2282]:
```suggestion
        adhoc = {}
```

File: odoo/models.py [hunk 7]
```diff
@@ -4490,7 +4569,8 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
         # the flush must be done before the _where_calc(), as the latter can do some selects
         self._flush_search(args, order=order)
 
-        query = self._where_calc(args)
+        adhoc = self._generate_adhoc_fields()
+        query = self._where_calc(args, adhoc=adhoc)
```
Comment (yelizariev) [inline odoo/models.py:4573]:
```suggestion
        query = self._where_calc(args)
```

File: odoo/models.py [hunk 8]
```diff
@@ -2225,6 +2238,9 @@ def _read_group_format_result(self, data, annotated_groupbys, groupby, domain):
         del data['id']
         return data
 
+    def _generate_adhoc_fields(self):
+        return {}
+
```
Comment (yelizariev) [inline odoo/models.py:2243]:
```suggestion
```

**Review by yelizariev — COMMENTED** [review]
If we delete ``_generate_adhoc_fields``, we'll get a solution for non stored related fields without further changes in API, but then we still need to solve ``read_progress_bar`` problem

**Comment by yelizariev** [comment]
Superseded with less general solutions:

read_progress_bar: PR v12 https://github.com/odoo/odoo/pull/67004
group by non stored related fields: PR master https://github.com/odoo/odoo/pull/66947

**Pull request closed without merging**